lightning/src/ln/channelmanager.rs

```diff
@@ -7906,6 +7704,346 @@ where
+) -> Result<(
+	onion_utils::Hop, [u8; 32], Option<(Result<PublicKey, secp256k1::Error>, u64, u64, u32)>
+), HTLCFailureMsg> 
+where
+	NS::Target: NodeSigner,
+	L::Target: Logger,
+{
+	macro_rules! return_malformed_err {
+		($msg: expr, $err_code: expr) => {
+			{
+				log_info!(logger, "Failed to accept/forward incoming HTLC: {}", $msg);
+				return Err(HTLCFailureMsg::Malformed(msgs::UpdateFailMalformedHTLC {
+					channel_id: msg.channel_id,
+					htlc_id: msg.htlc_id,
+					sha256_of_onion: Sha256::hash(&msg.onion_routing_packet.hop_data).into_inner(),
+					failure_code: $err_code,
+				}));
+			}
+		}
+	}
+
+	if let Err(_) = msg.onion_routing_packet.public_key {
+		return_malformed_err!("invalid ephemeral pubkey", 0x8000 | 0x4000 | 6);
+	}
+
+	let shared_secret = node_signer.ecdh(
+		Recipient::Node, &msg.onion_routing_packet.public_key.unwrap(), None
+	).unwrap().secret_bytes();
+
+	if msg.onion_routing_packet.version != 0 {
+		//TODO: Spec doesn't indicate if we should only hash hop_data here (and in other
+		//sha256_of_onion error data packets), or the entire onion_routing_packet. Either way,
+		//the hash doesn't really serve any purpose - in the case of hashing all data, the
+		//receiving node would have to brute force to figure out which version was put in the
+		//packet by the node that send us the message, in the case of hashing the hop_data, the
+		//node knows the HMAC matched, so they already know what is there...
+		return_malformed_err!("Unknown onion packet version", 0x8000 | 0x4000 | 4);
+	}
+	macro_rules! return_err {
+		($msg: expr, $err_code: expr, $data: expr) => {
+			{
+				log_info!(logger, "Failed to accept/forward incoming HTLC: {}", $msg);
+				return Err(HTLCFailureMsg::Relay(msgs::UpdateFailHTLC {
+					channel_id: msg.channel_id,
+					htlc_id: msg.htlc_id,
+					reason: HTLCFailReason::reason($err_code, $data.to_vec())
+						.get_encrypted_failure_packet(&shared_secret, &None),
+				}));
+			}
+		}
+	}
+
+	let next_hop = match onion_utils::decode_next_payment_hop(
+		shared_secret, &msg.onion_routing_packet.hop_data[..], msg.onion_routing_packet.hmac,
+		msg.payment_hash, node_signer
+	) {
+		Ok(res) => res,
+		Err(onion_utils::OnionDecodeErr::Malformed { err_msg, err_code }) => {
+			return_malformed_err!(err_msg, err_code);
+		},
+		Err(onion_utils::OnionDecodeErr::Relay { err_msg, err_code }) => {
+			return_err!(err_msg, err_code, &[0; 0]);
+		},
+	};
+
+	let next_packet_details = match next_hop {
+		onion_utils::Hop::Forward {
+			next_hop_data: msgs::InboundOnionPayload::Forward {
+				short_channel_id, amt_to_forward, outgoing_cltv_value
+			}, ..
+		} => {
+			let next_packet_pk = onion_utils::next_hop_pubkey(secp_ctx,
+				msg.onion_routing_packet.public_key.unwrap(), &shared_secret);
+			(next_packet_pk, short_channel_id, amt_to_forward, outgoing_cltv_value)
+		},
+		// We'll do receive checks in [`Self::construct_pending_htlc_info`] so we have access to the
+		// inbound channel's state.
```

Comment:
This comment doesn't really make sense anymore, can be removed in a separate commit.

Reply:
ok done